feat: Introduce Building Blocks page (#628)
meshcloud/meshcloud-docs: docs/administration.building-blocks.md

Next, you will have to enter how the implementation of the building block will work.

First of all, pick for which platform type(s) the building block is supported. Usually this is just one platform
but you could have one that works for multiple clouds or maybe it is even a cloud-agnostic block.

Next, you have to decide how often the building block can be booked for a meshTenant:

- Multiple: there is no limit to how many building blocks can exist for the meshTenant.
- Once: there can only be one building block for a single meshTenant. This is useful for core building blocks, e.g. a
  virtual network for a cloud tenant.

Next up is an important decision: the implementation type. As of now these are only two but we hope to add a lot more
in the future and even let you develop your own implementation types for building blocks. Now, you can either pick:

- Manual: this means the building block is fully managed manually. Ideally you always automate your building blocks
  but manual building blocks can be a good start to get familiar with the concept and give a clear ordering flow
  to your customers. It could also be an option to automate the building block in the future.

> - Terraform: this is the option to automate your building block. You will only need a Terraform file and meshStack will

Reviewer: ```suggestion
- Terraform: this is the option to automate your building block. You will only need a Terraform module and meshStack will
```

>   run this configuration everytime a user books a new building block. In a later step you can setup Inputs for your

Reviewer: ```suggestion
  run this configuration everytime a user adds a new building block to a meshTenant. In a later step you can setup Inputs for your
```

  Terraform script. There is also a [section](#developing--deploying-a-terraform-building-block)
  below with further tips for developing Terraform building blocks.

If you pick Terraform you have to enter a few further values:

> - Terraform Version: the version your Terraform files are based on.

Reviewer: ```suggestion
- Terraform Version: the version your Terraform module shall be executed with.
```

> - Git Repository URL: the URL of the Git repository where your Terraform files reside.

Reviewer: ```suggestion
- Git Repository URL: the URL of the Git repository where your Terraform module resides.
```

> - (Optional) Git Repository Path: the specific directory in the repository where the Terraform files reside.

Reviewer: ```suggestion
- (Optional) Git Repository Path: the specific directory in the repository where the Terraform module resides.
```

> - (Optional) Private SSH Key: upload an SSH key that has access to the provided Git repository.

Reviewer: ```suggestion
- (Optional) Private SSH Key: This is needed and the only option to access private repositories. Upload an SSH key that has access to the provided Git repository.
```

Lastly, you have to decide what the behavior of the building block is when it is deleted by the customer user. This can be

- Purge in meshStack: (this is the only option if you have a manual building block) The building block will be deleted
  in meshStack, but no underlying resources (e.g. in a cloud platform) will be deleted. You have to do this yourself.
- Delete Resources: this option will also clean up the underlying resources. The workings depend on the implementation

>   type, but e.g. for Terraform meshStack will call `terraform destroy` on the configuration files.

Reviewer: ```suggestion
  type, but e.g. for Terraform meshStack will call `terraform destroy` on the terraform module.
```

### Dependencies

> Please note that this the use of dependencies is completely optional and this step can be skipped.

> In some cases you might find that you have a multiple building block definitions, and one depends on the other.

Reviewer: ```suggestion
In some cases you might find that you have multiple building block definitions, and one depends on the other.
```

For example, you might have a Virtual Network, which is a prerequisite for a Virtual Machine.

You can model and enforce such dependencies in meshStack using the Building Blocks feature.

> In the wizard, on the "Dependencies" page, you have to select the building block(s) that the _about to be created_

Reviewer: ```suggestion
In the wizard, on the "Dependencies" page, you have to select the building block definition(s) that the _about to be created_
```

building block depends on. So in the example above you would have to do the following:

1. Create the Virtual Network Building Block without any dependencies.
2. Create the Virtual Machine Building Block and select the Virtual Network as a dependency.

> > For Terraform users: you can provide the output of the Virtual Network Building Block as input for the Virtual

Reviewer: d: We should somewhere very prominently mention that any kind of Building Block can be connected to any other. This can be achieved by using outputs of one block definition as input in another block definition. I think this is the super string integration feature of our building blocks. You can just connect all kinds of blocks, independent of their implementation type

> Machine building block.

### Building Block Inputs

As with a lot of configuration, input is required. Building Blocks in meshStack can collect as much input as you want,
in many ways possible.

There are different kinds of inputs, with different behavior, which can be decided under the "Source" section. You can

> read in the meshPanel what each kind of input does and how it behaves.

Reviewer: d: I think it would also be worth mentioning those typoes here for readers of the docs who are just inetrested in meshStack, but don't use it yet. With all those input types we can create some wow moments on what is possible here

The inputs' behavior can be further decided:

- Provide as: you can decide whether the input is regularly provided or as an environment variable.
  (only applies to Terraform Building Blocks)
- Sensitivity: you can decide that an input is sensitive (e.g. a secret). It will then be encrypted in meshStack and not

>   shown to any users.

Reviewer: ```suggestion
  shown to any users. Also meshStack only has very limited access to this secret. Only once Terraform gets executed, the secret is decrypted.
```

> ### Building Block Outputs (Terraform-only)

Reviewer: d: We actually also have outputs for manual blocks. You just cannot define them explicitly. You just get all inputs as outputs, so you can use those outputs as inputs for another block definition

You can also extract the Terraform outputs into meshStack. This is useful in two scenarios:

1. You want to provide an output variable to a dependent building block. As the example that was mentioned above, if

>    you have a virtual machine that depends on the IP Address (? - check) of a virtual network, you can output the

Reviewer: ```suggestion
   you have a virtual machine that depends on the VNET ID of a virtual network, you can output the
```

   variable and provide it as an input to the dependent virtual machine building block.
2. You want the customer user to be able to see the output result. This might be useful if there is an output that is
   important to the customer user. They will be able to read the value in their tenant control plane.

Last but not least, save your building block definition, and it will be ready to be consumed by your customer users!

## Developing & Deploying a Terraform Building Block

For Terraform blocks, meshStack will run the required Terraform in the background.

> ### How to debug

Reviewer: d: I think at least with a next PR we should also mention the option to rerun a building block (e.g. if it failed and the operator now fixed the underlying issue). And we should mention that blocks can also be deleted

You can view the logs of meshStack running the Terraform files. To do so, go to the Tenant Overview in the Admin Area
and look for a meshTenant that has booked the building block. In there you will be able to find the building block
and read the logs that were written while executing the Terraform.

> ## Editing Building Blocks

Reviewer: ```suggestion
## Editing Building Block Definitions
```
